File: modules/importer.py
# import python modules
# *************************************************************************
# *                                                                       *
# * OpenAdmin                                                             *
# * Copyright (c) OpenPanel. All Rights Reserved.                         *
# * Version: 1.6.0                                                        *
# * Build Date: 2025-09-23 11:25:17                                       *
# *                                                                       *
# *************************************************************************
# *                                                                       *
# * Email: info@openpanel.com                                             *
# * Website: https://openpanel.com                                        *
# *                                                                       *
# *************************************************************************
# *                                                                       *
# * This software is furnished under a license and may be used and copied *
# * only  in  accordance  with  the  terms  of such  license and with the *
# * inclusion of the above copyright notice.  This software  or any other *
# * copies thereof may not be provided or otherwise made available to any *
# * other person.  No title to and  ownership of the software is  hereby *
# * transferred.                                                          *
# *                                                                       *
# * You may not reverse  engineer, decompile, defeat  license  encryption *
# * mechanisms, or  disassemble this software product or software product *
# * license.  OpenPanel may terminate this license if you don't comply    *
# * with any of the terms and conditions set forth in our end user        *
# * license agreement (EULA).  In such event,  licensee  agrees to return *
# * licensor  or destroy  all copies of software  upon termination of the *
# * license.                                                              *
# *                                                                       *
# * Please see the EULA file for the full End User License Agreement.     *
# *                                                                       *
# *************************************************************************
import os
import re
import json
from flask import Flask, Response, render_template, request, g, jsonify, session, redirect, url_for, flash, get_flashed_messages
import subprocess
import datetime
from collections import OrderedDict
import psutil
from subprocess import run, Popen, PIPE
import requests
from requests.exceptions import RequestException, HTTPError, ConnectionError, Timeout
import glob
import signal

# import our modules
from app import app, admin_required_route, load_openpanel_config
from modules.helpers import get_all_users, get_user_and_plan_count, get_all_plans

# CPANEL 2 OPENPANEL IMPORT
def is_pid_running(pid):
    print(f"IMPORTER - Check if a process with the given PID is running: {pid}")        
    try:
        subprocess.check_output(['ps', '-p', str(pid)])
        return True
    except subprocess.CalledProcessError:
        return False





@app.route('/user/import', methods=['GET'])
@admin_required_route
def import_cp_user():
    current_route = request.path
    hosting_plans = get_all_plans()
    plans = hosting_plans if hosting_plans else []
    messages = get_flashed_messages()       
    return render_template('users/import/import.html', 
        title='Import User', 
        messages=messages, 
        plans=plans, 
        app=app, 
        current_route=current_route
        )











@app.route('/import/cpanel', methods=['GET', 'POST'])
@admin_required_route
def import_users_from_cpanel():

    if request.method == 'POST':
        backup_path = request.form.get('path')
        plan_name = request.form.get('plan_name')
        print(f"IMPORTER - Importing user account from cPanel backup: {backup_path} to plan: {plan_name}")

        # Validate inputs
        if not backup_path or not plan_name:
            print(f"IMPORTER - Aborting import: path or plan_name are not provided.")        
            flash('Backup path and plan name are required.', 'danger')
            return redirect(url_for('import_cp_user'))
            
        temp_dir = '/tmp/cPanel-to-OpenPanel/'
        repo_url = 'https://github.com/stefanpejcic/cPanel-to-OpenPanel'
        clone_command = ['git', 'clone', repo_url, temp_dir]
        import_script = os.path.join(temp_dir, 'cp-import.sh')
        import_command = ['bash', import_script, '--backup-location', backup_path, '--plan-name', plan_name]

        # Remove existing directory
        if os.path.exists(temp_dir):
            print(f"IMPORTER - Removing: {temp_dir}")
            subprocess.run(['rm', '-rf', temp_dir], check=True)

        try:
            print(f"IMPORTER - Executing: {clone_command}")
            subprocess.run(clone_command, check=True)
            subprocess.Popen(import_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            flash(f'Import process has started.', 'success')
        except subprocess.CalledProcessError as e:
            print(f"IMPORTER - Error executing command, error: {str(e)}")
            flash(f'Error during execution: {e.stderr}', 'warning')
            return redirect(url_for('import_cp_user'))
        except Exception as e:
            flash(str(e), 'danger')
            print(f"IMPORTER - Error executing command, executing: {str(e)}")
            return redirect(url_for('import_cp_user'))     
        return redirect(url_for('import_users_from_cpanel'))  

    # after post and on get
    log_dir = '/var/log/openpanel/admin/imports/'
    log_files = []
    statuses = {}
    now = datetime.datetime.now()
    print(f"IMPORTER - Checking statuses for imports in: {log_dir}")

    try:
        files = [f for f in os.listdir(log_dir) if f.endswith('.log')]
    except (FileNotFoundError, PermissionError):
        files = []

    for log_file in files:
        log_file_path = os.path.join(log_dir, log_file)
        status = "unknown"
        try:
            with open(log_file_path, 'r') as file:
                lines = file.readlines()
                if lines:
                    # Read the first two lines
                    first_line = lines[0].strip()
                    pid_line = lines[1].strip() if len(lines) > 1 else ""
                    
                    # Extract PID from the second line if present
                    pid = None
                    if "PID:" in pid_line:
                        try:
                            pid = int(pid_line.split('PID:')[1].strip())
                        except ValueError:
                            pid = None
                            
                    # Determine status based on the PID
                    if pid and is_pid_running(pid):
                        status = "running"
                    else:
                        # Check the last line for success or failure
                        last_line = lines[-1].strip()
                        if "SUCCESS:" in last_line:
                            status = "completed"
                        elif "FATAL ERROR:" in last_line:
                            status = "failed"
                        else:
                            status = "unknown"

        except (FileNotFoundError, PermissionError):
            status = "unknown"
        
        log_files.append({
            'filename': log_file,
            'status': status
        })

    return render_template('users/import/list_import_cp_logs.html', title='cPanel Imports', log_files=log_files)




@app.route('/import/user/log/<path:log_filename>', methods=['GET'])
@admin_required_route
def view_transfer_import_log(log_filename):
    try:
        log_dir = '/var/log/openpanel/admin/transfers/'
        log_path = os.path.join(log_dir, log_filename)

        if not os.path.isfile(log_path):
            print(f"IMPORTER - Log file: {log_path} does not exist.")
            if request.args.get('output') == 'json':
                return jsonify({
                    'status': 'error',
                    'message': 'Log file does not exist.'
                }), 404
            return Response("Log file does not exist.", status=404, mimetype='text/plain')

        print(f"IMPORTER - Reading log file: {log_path}")
        with open(log_path, 'r') as file:
            log_content = file.read()

        if request.args.get('output') == 'json':
            return jsonify({
                'status': 'success',
                'log': log_content
            })

        return Response(log_content, mimetype='text/plain')

    except OSError as e:
        print(f"IMPORTER - Error accessing log file: {str(e)")
        if request.args.get('output') == 'json':
            return jsonify({
                'status': 'error',
                'message': f'Error accessing log file: {str(e)}'
            }), 500
        return Response(f"Error accessing log file: {str(e)}", status=500, mimetype='text/plain')

    except Exception as e:
        print(f"IMPORTER - Exception reading log file: {str(e)")
        if request.args.get('output') == 'json':
            return jsonify({
                'status': 'error',
                'message': str(e)
            }), 500
        return Response(str(e), status=500, mimetype='text/plain')


@app.route('/import/cpanel/log/<path:log_filename>', methods=['GET'])
@admin_required_route
def view_cp_import_log(log_filename):
    try:
        log_dir = '/var/log/openpanel/admin/imports/'
        log_path = os.path.join(log_dir, log_filename)

        if not os.path.isfile(log_path):
            print(f"IMPORTER - Log does not exist: {log_path}")
            if request.args.get('output') == 'json':
                return jsonify({
                    'status': 'error',
                    'message': 'Log file does not exist.'
                }), 404
            return Response("Log file does not exist.", status=404, mimetype='text/plain')

        print(f"IMPORTER - Reading: {log_path}")
        with open(log_path, 'r') as file:
            log_content = file.read()

        if request.args.get('output') == 'json':
            return jsonify({
                'status': 'success',
                'log': log_content
            })

        return Response(log_content, mimetype='text/plain')

    except OSError as e:
        print(f"IMPORTER - Error accessing log file: {str(e)}")
        if request.args.get('output') == 'json':
            return jsonify({
                'status': 'error',
                'message': f'Error accessing log file: {str(e)}'
            }), 500
        return Response(f"Error accessing log file: {str(e)}", status=500, mimetype='text/plain')

    except Exception as e:
        print(f"IMPORTER - Exception accessing log file: {str(e)}")
        if request.args.get('output') == 'json':
            return jsonify({
                'status': 'error',
                'message': str(e)
            }), 500
        return Response(str(e), status=500, mimetype='text/plain')





# https://github.com/stefanpejcic/OpenPanel/issues/239
@app.route('/json/cpanel-backups', methods=['GET'])
@admin_required_route
def list_cpanel_backups():
    search_dirs = ['/', '/home', '/root']
    pattern = 'backup-*.tar.gz'
    print(f"IMPORTER - Checking for backup-*.tar.gz files in paths: / /home /root")
    backups = []
    for directory in search_dirs:
        full_pattern = os.path.join(directory, pattern)
        found_files = glob.glob(full_pattern)
        backups.extend(found_files)
    return jsonify(backups)


@app.route('/json/transfers', methods=['GET'])
@admin_required_route
def list_transfers():
    search_dir = '/var/log/openpanel/admin/transfers/'
    pattern = '*.log'
    full_pattern = os.path.join(search_dir, pattern)
    print(f"IMPORTER - Listing transfers: {full_pattern}")
    files = glob.glob(full_pattern)
    return jsonify(files)




@app.route('/json/transfers/<username>', methods=['GET'])
@admin_required_route
def list_transfers_for(username):
    if "SUSPENDED_" in username:
        username = username.rsplit("_", 1)[-1]

    search_dir = '/var/log/openpanel/admin/transfers/'
    pattern = f'{username}_*.log'
    full_pattern = os.path.join(search_dir, pattern)
    files = glob.glob(full_pattern)

    transfers = []
    print(f"IMPORTER - Listing transfer logs: {full_pattern}")

    for filepath in files:
        transfer = {
            'filename': os.path.basename(filepath),
            'file': filepath,
            'status': 'unknown',
            'pid': None
        }

        try:
            with open(filepath, 'r') as f:
                lines = f.readlines()

            pid_line = next((line for line in lines if 'PID:' in line), None)
            if pid_line:
                pid = int(pid_line.strip().split('PID:')[1])
                transfer['pid'] = pid

                try:
                    os.kill(pid, 0)
                    transfer['status'] = 'in progress'
                except ProcessLookupError:
                    if lines and 'SUCCESS:' in lines[-1]:
                        transfer['status'] = 'success'
                    else:
                        transfer['status'] = 'failed'
            else:
                transfer['status'] = 'failed'

        except Exception as e:
            transfer['error'] = str(e)
            transfer['status'] = 'error'

        transfers.append(transfer)

    return jsonify(transfers)


# OPENPANEL 2 OPENPANEL IMPORT


def is_domain(server):
    print(f"IMPORTER - Validating domain: {server}")
    domain_regex = re.compile(
        r'^(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$'
    )
    return domain_regex.match(server) is not None

def configure_iptables(server):
    print(f"IMPORTER - Executing: csf -ta {server} 3600")
    try:
        subprocess.run(['csf', '-ta', server, '3600'], check=True)
        return True
    except subprocess.CalledProcessError as e:
        print(f"IMPORTER - Error temporary iptables: {e}")
        return False


def mask_secret(s: str, show=2, mask_char='*'):
    if not s:
        return None
    visible = s[-show:] if len(s) > show else s
    masked_len = max(1, len(s) - len(visible))
    return mask_char * masked_len + visible

@app.route('/import/transfer/', methods=['GET', 'POST'])
@admin_required_route
def import_users():
    if request.method == 'GET':
        log_dir = '/var/log/openpanel/admin/transfers/'
        log_files = []
        statuses = {}
        now = datetime.datetime.now()

        print(f"IMPORTER - Checking for .log files in: {log_dir}")
        try:
            files = [f for f in os.listdir(log_dir) if f.endswith('.log')]
        except (FileNotFoundError, PermissionError):
            files = []

        for log_file in files:
            log_file_path = os.path.join(log_dir, log_file)
            status = "unknown"
            try:
                with open(log_file_path, 'r') as file:
                    lines = file.readlines()
                    if lines:
                        # Read the first two lines
                        first_line = lines[0].strip()
                        pid_line = lines[1].strip() if len(lines) > 1 else ""
                        
                        # Extract PID from the second line if present
                        pid = None
                        if "PID:" in pid_line:
                            try:
                                pid = int(pid_line.split('PID:')[1].strip())
                            except ValueError:
                                pid = None
                                
                        # Determine status based on the PID
                        if pid and is_pid_running(pid):
                            status = "running"
                        else:
                            # Check the last line for success or failure
                            last_line = lines[-1].strip()
                            if "SUCCESS:" in last_line:
                                status = "completed"
                            elif "FATAL ERROR:" in last_line:
                                status = "failed"
                            else:
                                status = "unknown"

            except (FileNotFoundError, PermissionError):
                status = "unknown"
            
            log_files.append({
                'filename': log_file,
                'status': status
            })

        return render_template('users/import/import_users.html', title='Account Transfer Logs', log_files=log_files)


    elif request.method == 'POST':
        openpanel_username = request.form.get('openpanel_username')
        server = request.form.get('server')
        username = request.form.get('username', 'root')
        password = request.form.get('password')
        port = request.form.get('port')
        live_transfer = request.form.get('live_transfer')

        print(f"IMPORTER - Received transfer request for: username: {openpanel_username} | "
            f"live_transfer: {live_transfer} | server: {server} | ssh_port: {port} | "
            f"ssh_username: {username} | ssh_password: {mask_secret(password, show=2)}")
        if not openpanel_username or not server:
            print(f"IMPORTER - Aborting transfer: server IP or username are not provided.")
            flash(f"Error: Server IP and OpenPanel username are required.", 'danger')
            return redirect(f'/users/{openpanel_username}#transfer')

        try:
            local_command = [
                "opencli",
                "user-transfer",
                "--account", openpanel_username,
                "--host", server,
                "--username", username
            ]

            if password:
                local_command += ["--password", password]

            if port:
                local_command += ["--port", str(port)]

            if live_transfer:
                local_command.append("--live-transfer")

            if not is_domain(server):
                configure_iptables(server)

            print(f"IMPORTER - Executing: {local_command}")
            subprocess.Popen(
                local_command,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                stdin=subprocess.DEVNULL,
                start_new_session=True
            )

            flash(f"Transfer process for account {openpanel_username} started in the background.", "success")

        except Exception as e:
            msg = f"Error starting transfer: {e}"
            print(f"IMPORTER - {msg}")
            flash(msg, "danger")

        return redirect(f'/users/{openpanel_username}#transfer')

File: internal/handlers/importer.go
// This file implements cPanel/CyberPanel/OpenPanel backup import and
// OpenPanel-to-OpenPanel account transfer tooling. This is an
// Enterprise-only feature, but no Enterprise gate is added here since no
// other module's routes are gated that way either -- see the license
// package's RequireEnterprise if that needs wiring in later.
package handlers

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"

	"github.com/gorilla/csrf"

	"openadmin/internal/auth"
	"openadmin/internal/paneldb"
	"openadmin/internal/webtemplates"
)

// Importer bundles the /user/import, /import/*, and /json/{backup-files,transfers}
// handlers.
type Importer struct {
	Sessions *auth.Manager
	MySQL    *sql.DB
}

// ImporterOpenPanelImportLogDir / ImporterTransferLogDir / ImporterRestoreTempDir
// are the hardcoded paths used by the import/transfer tooling.
var (
	ImporterOpenPanelImportLogDir = "/var/log/openpanel/admin/imports/"
	ImporterTransferLogDir        = "/var/log/openpanel/admin/transfers/"
	ImporterRestoreTempDir        = "/home/temprestoreditforopbackup"
)

// isPidRunning is true only if `ps -p PID` exits zero. Any failure
// (including ps itself being unavailable) is treated as "not running".
func isPidRunning(pid int) bool {
	return exec.Command("ps", "-p", strconv.Itoa(pid)).Run() == nil
}

type importLogStatus struct {
	Filename string `json:"filename"`
	Status   string `json:"status"`
}

// determineLogStatus is the log-status logic shared by
// ServeImportFromBackup and ServeImportTransfer's GET branches.
func determineLogStatus(path string) string {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "unknown"
	}
	lines := strings.Split(string(raw), "\n")
	if len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}
	if len(lines) == 0 {
		return "unknown"
	}

	pidLine := ""
	if len(lines) > 1 {
		pidLine = strings.TrimSpace(lines[1])
	}
	pid, havePID := 0, false
	if idx := strings.Index(pidLine, "PID:"); idx != -1 {
		if n, err := strconv.Atoi(strings.TrimSpace(pidLine[idx+len("PID:"):])); err == nil {
			pid, havePID = n, true
		}
	}

	if havePID && isPidRunning(pid) {
		return "running"
	}

	lastLine := strings.TrimSpace(lines[len(lines)-1])
	switch {
	case strings.Contains(lastLine, "SUCCESS:"):
		return "completed"
	case strings.Contains(lastLine, "FATAL ERROR:"):
		return "failed"
	default:
		return "unknown"
	}
}

func listLogFilesWithStatus(logDir string) []importLogStatus {
	entries, err := os.ReadDir(logDir)
	if err != nil {
		return []importLogStatus{}
	}
	result := []importLogStatus{}
	for _, e := range entries {
		name := e.Name()
		if e.IsDir() || !strings.HasSuffix(name, ".log") {
			continue
		}
		result = append(result, importLogStatus{Filename: name, Status: determineLogStatus(filepath.Join(logDir, name))})
	}
	return result
}

// ServeImportUser handles GET /user/import.
func (im *Importer) ServeImportUser(w http.ResponseWriter, r *http.Request) {
	plans, _ := paneldb.GetAllPlans(im.MySQL, nil)
	webtemplates.Render(w, "users_import_import.html", mergeChrome(map[string]interface{}{
		"Plans":     plans,
		"CSRFToken": csrf.Token(r),
		"Flashes":   auth.PopFlashes(w, r, im.Sessions),
	}, r, "Import User"))
}

var importerBackupPanelDisplayNames = map[string]string{
	"cpanel":     "cPanel",
	"cyberpanel": "CyberPanel",
	"openpanel":  "OpenPanel",
}

// importerRestoreOpenPanelBackupRun starts a fire-and-forget background
// restore for the openpanel backup type.
var importerRestoreOpenPanelBackupRun = func(backupPath string) error {
	return exec.Command("opencli", "user-restore", "--file", backupPath,
		"--temp-dir="+ImporterRestoreTempDir).Start()
}

// importerCloneAndRunImportScriptRun handles the cpanel/cyberpanel case:
// removes any stale clone dir, clones the panel-specific import repo (a
// clone failure is reported distinctly, under a "warning" category with
// the git stderr), then fires off the import script in the background.
var importerCloneAndRunImportScriptRun = func(displayName, backupPath, planName string) (cloneFailed bool, err error) {
	tempDir := "/tmp/" + displayName + "-to-OpenPanel/"
	repoURL := "https://github.com/stefanpejcic/" + displayName + "-to-OpenPanel"
	importScript := filepath.Join(tempDir, "cp-import.sh")

	if _, statErr := os.Stat(tempDir); statErr == nil {
		exec.Command("rm", "-rf", tempDir).Run()
	}

	var stderr strings.Builder
	cloneCmd := exec.Command("git", "clone", repoURL, tempDir)
	cloneCmd.Stderr = &stderr
	if err := cloneCmd.Run(); err != nil {
		return true, errors.New(stderr.String())
	}

	importCmd := exec.Command("bash", importScript,
		fmt.Sprintf("--backup-location='%s'", backupPath),
		fmt.Sprintf("--plan-name='%s'", planName))
	if err := importCmd.Start(); err != nil {
		return false, err
	}
	return false, nil
}

// ServeImportFromBackup handles GET/POST /import/{panel_type}.
func (im *Importer) ServeImportFromBackup(w http.ResponseWriter, r *http.Request) {
	panelType := strings.ToLower(r.PathValue("panel_type"))
	displayName, ok := importerBackupPanelDisplayNames[panelType]
	if !ok {
		writeJSONError(w, http.StatusBadRequest, "Unsupported panel type: "+panelType)
		return
	}

	if r.Method == http.MethodPost {
		r.ParseForm()
		backupPath := r.PostFormValue("path")
		planName := r.PostFormValue("plan_name")

		if panelType == "openpanel" {
			if backupPath == "" {
				auth.AddFlash(w, r, im.Sessions, "Backup file is required.", "danger")
				http.Redirect(w, r, "/user/import", http.StatusSeeOther)
				return
			}
			if err := importerRestoreOpenPanelBackupRun(backupPath); err != nil {
				auth.AddFlash(w, r, im.Sessions, err.Error(), "danger")
				http.Redirect(w, r, "/user/import", http.StatusSeeOther)
				return
			}
			auth.AddFlash(w, r, im.Sessions, fmt.Sprintf("Import process from %s (OpenPanel backup) has started.", backupPath), "success")
			http.Redirect(w, r, "/import/"+panelType, http.StatusSeeOther)
			return
		}

		if backupPath == "" || planName == "" {
			auth.AddFlash(w, r, im.Sessions, "Backup path and plan name are required.", "danger")
			http.Redirect(w, r, "/user/import", http.StatusSeeOther)
			return
		}

		cloneFailed, err := importerCloneAndRunImportScriptRun(displayName, backupPath, planName)
		if err != nil {
			if cloneFailed {
				auth.AddFlash(w, r, im.Sessions, "Error during execution: "+err.Error(), "warning")
			} else {
				auth.AddFlash(w, r, im.Sessions, err.Error(), "danger")
			}
			http.Redirect(w, r, "/user/import", http.StatusSeeOther)
			return
		}
		auth.AddFlash(w, r, im.Sessions, fmt.Sprintf("Import process from %s (%s backup) has started.", backupPath, displayName), "success")
		http.Redirect(w, r, "/import/"+panelType, http.StatusSeeOther)
		return
	}

	webtemplates.Render(w, "users_import_list_import_logs.html", mergeChrome(map[string]interface{}{
		"LogFiles": listLogFilesWithStatus(ImporterOpenPanelImportLogDir),
		"Flashes":  auth.PopFlashes(w, r, im.Sessions),
	}, r, "Account Imports"))
}

// serveImportLog is shared by ServeViewTransferImportLog and
// ServeViewAccountImportLog, which are otherwise identical aside from
// which log directory they read from.
func serveImportLog(w http.ResponseWriter, r *http.Request, logDir string) {
	logFilename := r.PathValue("log_filename")
	wantJSON := r.URL.Query().Get("output") == "json"

	logPath, ok := safeJoinOr400(logDir, logFilename)
	if !ok {
		if wantJSON {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusBadRequest)
			json.NewEncoder(w).Encode(map[string]string{"status": "error", "message": "Invalid log filename."})
			return
		}
		http.Error(w, "Invalid log filename.", http.StatusBadRequest)
		return
	}

	info, err := os.Stat(logPath)
	if err != nil || info.IsDir() {
		if wantJSON {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusNotFound)
			json.NewEncoder(w).Encode(map[string]string{"status": "error", "message": "Log file does not exist."})
			return
		}
		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusNotFound)
		w.Write([]byte("Log file does not exist."))
		return
	}

	content, err := os.ReadFile(logPath)
	if err != nil {
		if wantJSON {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusInternalServerError)
			json.NewEncoder(w).Encode(map[string]string{"status": "error", "message": "Error accessing log file: " + err.Error()})
			return
		}
		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusInternalServerError)
		w.Write([]byte("Error accessing log file: " + err.Error()))
		return
	}

	if wantJSON {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]string{"status": "success", "log": string(content)})
		return
	}
	w.Header().Set("Content-Type", "text/plain")
	w.Write(content)
}

// ServeViewTransferImportLog handles GET /import/user/log/{log_filename}.
func (im *Importer) ServeViewTransferImportLog(w http.ResponseWriter, r *http.Request) {
	serveImportLog(w, r, ImporterTransferLogDir)
}

// ServeViewAccountImportLog handles GET /import/account/log/{log_filename}.
func (im *Importer) ServeViewAccountImportLog(w http.ResponseWriter, r *http.Request) {
	serveImportLog(w, r, ImporterOpenPanelImportLogDir)
}

var importerBackupSearchDirs = []string{"/", "/home", "/root"}
var importerBackupPatterns = []string{`^backup-.*\.tar\.gz$`, `^.*_\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2}\.tar\.gz$`}

// ServeListBackupFiles handles GET /json/backup-files: a shallow
// (non-recursive) scan of /, /home, and /root for cPanel/CyberPanel
// (backup-*.tar.gz) or OpenPanel (<user>_<date>_<time>.tar.gz) backups.
func (im *Importer) ServeListBackupFiles(w http.ResponseWriter, r *http.Request) {
	var backupRes []*regexp.Regexp
	for _, p := range importerBackupPatterns {
		backupRes = append(backupRes, regexp.MustCompile(p))
	}

	seen := map[string]bool{}
	backups := []string{}
	for _, dir := range importerBackupSearchDirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		for _, e := range entries {
			if e.IsDir() {
				continue
			}
			name := e.Name()
			matched := false
			for _, re := range backupRes {
				if re.MatchString(name) {
					matched = true
					break
				}
			}
			if !matched {
				continue
			}
			full := filepath.Join(dir, name)
			if !seen[full] {
				seen[full] = true
				backups = append(backups, full)
			}
		}
	}
	writeJSON(w, backups)
}

// ServeListTransfers handles GET /json/transfers: every *.log file
// directly under the transfers dir.
func (im *Importer) ServeListTransfers(w http.ResponseWriter, r *http.Request) {
	matches, _ := filepath.Glob(filepath.Join(ImporterTransferLogDir, "*.log"))
	if matches == nil {
		matches = []string{}
	}
	writeJSON(w, matches)
}

type transferStatus struct {
	Filename string `json:"filename"`
	File     string `json:"file"`
	Status   string `json:"status"`
	PID      *int   `json:"pid"`
	Error    string `json:"error,omitempty"`
}

// importerPidAlive has a 3-way outcome: nil (alive, signalable), ESRCH
// ("no such process"), or any other error (e.g. EPERM: the process exists
// but belongs to another user). Only ESRCH is treated as "not running"
// below; any other error falls through to t.Status = "error" instead.
// Injectable for tests.
var importerPidAlive = func(pid int) error {
	return syscall.Kill(pid, 0)
}

// scanTransfersForUsername scans the transfer log directory for logs
// belonging to username and reports each one's current status. Shared by
// the HTML transfers-for-user route and its JSON API counterpart -- the
// two differ only in how they authorize the caller, not in how they read
// the logs.
func scanTransfersForUsername(username string) []transferStatus {
	matches, _ := filepath.Glob(filepath.Join(ImporterTransferLogDir, username+"_*.log"))
	transfers := make([]transferStatus, 0, len(matches))
	for _, path := range matches {
		t := transferStatus{Filename: filepath.Base(path), File: path, Status: "unknown"}

		raw, err := os.ReadFile(path)
		if err != nil {
			t.Status = "error"
			t.Error = err.Error()
			transfers = append(transfers, t)
			continue
		}
		lines := strings.Split(string(raw), "\n")
		if len(lines) > 0 && lines[len(lines)-1] == "" {
			lines = lines[:len(lines)-1]
		}

		var pidLine string
		for _, l := range lines {
			if strings.Contains(l, "PID:") {
				pidLine = l
				break
			}
		}

		if pidLine == "" {
			t.Status = "failed"
			transfers = append(transfers, t)
			continue
		}

		idx := strings.Index(pidLine, "PID:")
		pid, perr := strconv.Atoi(strings.TrimSpace(pidLine[idx+len("PID:"):]))
		if perr != nil {
			t.Status = "error"
			t.Error = perr.Error()
			transfers = append(transfers, t)
			continue
		}
		t.PID = &pid

		switch err := importerPidAlive(pid); {
		case err == nil:
			t.Status = "in progress"
		case errors.Is(err, syscall.ESRCH):
			if len(lines) > 0 && strings.Contains(lines[len(lines)-1], "SUCCESS:") {
				t.Status = "success"
			} else {
				t.Status = "failed"
			}
		default:
			t.Status = "error"
			t.Error = err.Error()
		}
		transfers = append(transfers, t)
	}
	return transfers
}

// ServeListTransfersFor handles GET /json/transfers/{username}.
func (im *Importer) ServeListTransfersFor(w http.ResponseWriter, r *http.Request) {
	username := r.PathValue("username")
	if strings.Contains(username, "SUSPENDED_") {
		if idx := strings.LastIndex(username, "_"); idx != -1 {
			username = username[idx+1:]
		}
	}

	cu := auth.CurrentUser(r)
	actingUsername, actingRole := "", ""
	if cu != nil {
		actingUsername, actingRole = cu.Username, cu.Role
	}
	if !paneldb.CheckIfOwnerForUser(im.MySQL, username, actingUsername, actingRole) {
		http.Error(w, "Forbidden", http.StatusForbidden)
		return
	}

	writeJSON(w, scanTransfersForUsername(username))
}

var importerDomainRe = regexp.MustCompile(`^(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}$`)

func isDomain(server string) bool {
	return importerDomainRe.MatchString(server)
}

// configureIptablesRun sets a temporary ConfigServer Firewall allow rule
// for an IP-addressed (non-domain) transfer target.
var configureIptablesRun = func(server string) bool {
	return exec.Command("csf", "-ta", server, "3600").Run() == nil
}

// importerStartTransferRun fires off `opencli user-transfer` in the
// background without waiting for it to complete.
var importerStartTransferRun = func(args []string) error {
	return exec.Command(args[0], args[1:]...).Start()
}

// ServeImportTransfer handles GET/POST /import/transfer/.
func (im *Importer) ServeImportTransfer(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodPost {
		r.ParseForm()
		openpanelUsername := r.PostFormValue("openpanel_username")
		server := r.PostFormValue("server")
		username := r.PostFormValue("username")
		if username == "" {
			username = "root"
		}
		password := r.PostFormValue("password")
		port := r.PostFormValue("port")
		liveTransfer := r.PostFormValue("live_transfer")

		if openpanelUsername == "" || server == "" {
			auth.AddFlash(w, r, im.Sessions, "Error: Server IP and OpenPanel username are required.", "danger")
			http.Redirect(w, r, "/users/"+openpanelUsername+"#transfer", http.StatusSeeOther)
			return
		}

		args := []string{"opencli", "user-transfer", "--account", openpanelUsername, "--host", server, "--username", username}
		if password != "" {
			args = append(args, "--password", password)
		}
		if port != "" {
			args = append(args, "--port", port)
		}
		if liveTransfer != "" {
			args = append(args, "--live-transfer")
		}

		if !isDomain(server) {
			configureIptablesRun(server)
		}

		if err := importerStartTransferRun(args); err != nil {
			auth.AddFlash(w, r, im.Sessions, "Error starting transfer: "+err.Error(), "danger")
		} else {
			auth.AddFlash(w, r, im.Sessions, fmt.Sprintf("Transfer process for account %s started in the background.", openpanelUsername), "success")
		}
		http.Redirect(w, r, "/users/"+openpanelUsername+"#transfer", http.StatusSeeOther)
		return
	}

	webtemplates.Render(w, "users_import_import_users.html", mergeChrome(map[string]interface{}{
		"LogFiles": listLogFilesWithStatus(ImporterTransferLogDir),
		"Flashes":  auth.PopFlashes(w, r, im.Sessions),
	}, r, "Account Transfer Logs"))
}
